common/src/main/java/com/google/android/fhir/MoreTypes.kt

```diff
@@ -40,6 +41,7 @@ fun equals(a: Type, b: Type): Boolean {
   // Codes with the same system and code values are considered equal even if they have different
   // display values.
   if (a is Coding && b is Coding) return a.system == b.system && a.code == b.code
+  if (a is Reference && b is Reference) return a.reference == b.reference
 
   throw NotImplementedError("Comparison for type ${a::class.java} not supported.")
```

Comment:
Maybe consider using 
```suggestion
when(a) {
  is Coding -> a.system == b.system && a.code == b.code
  is Reference -> a.reference == b.reference
  else -> throw NotImplementedError("Comparison for type ${a::class.java} not supported.")
}
```

Reply:
Could you please verify whether the properties of 'b' are resolved, and no build errors occur with this change? It appears that checking the type of the instance for 'b' before accessing its properties is necessary on my end.

Reply:
Oh yeah Kotlin might not be detecting the check `a::class != b::class`.

Try adding this check inside when ?

Reply:
```suggestion
  return a.run {
    if (this::class != b::class) false
    else if (this === b) true
    else when (this) {
      is PrimitiveType<*> -> primitiveValue() == b.primitiveValue()
      is Coding -> b is Coding && system == b.system && code == b.code   // Assumes b is also a Coding
      is Reference -> b is Reference && reference == b.reference         // Assumes b is also a Reference
      else -> throw NotImplementedError("Comparison for type ${this::class.java} not supported.")
    }
  }
```

^This works

Reply:
Acknowledged.